Sources/ShopifyCheckoutKit/AnalyticsEvents.swift

```diff
@@ -0,0 +1,3579 @@
+    let id: String?
+    /// Arbitrary name of the custom event
+    let name: String?
+    /// The timestamp of when the customer event occurred, in [ISO
+    /// 8601](https://en.wikipedia.org/wiki/ISO_8601) format
+    let timestamp: String?
+
+    enum CodingKeys: String, CodingKey {
+        case context, customData, id, name, timestamp
+    }
+}
+
+// MARK: CustomEvent convenience initializers and mutators
+
+extension CustomEvent {
+    init(data: Data) throws {
+        self = try newJSONDecoder().decode(CustomEvent.self, from: data)
+    }
+
+    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
+        guard let data = json.data(using: encoding) else {
+            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
+        }
+        try self.init(data: data)
+    }
+
+    init(fromURL url: URL) throws {
+        try self.init(data: try Data(contentsOf: url))
+    }
+
+    func with(
+        context: Context?? = nil,
+        customData: CustomData?? = nil,
+        id: String?? = nil,
+        name: String?? = nil,
+        timestamp: String?? = nil
+    ) -> CustomEvent {
+        return CustomEvent(
+            context: context ?? self.context,
+            customData: customData ?? self.customData,
+            id: id ?? self.id,
+            name: name ?? self.name,
+            timestamp: timestamp ?? self.timestamp
+        )
+    }
+
+    func jsonData() throws -> Data {
+        return try newJSONEncoder().encode(self)
+    }
+
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// A free-form object representing data specific to a custom event provided by
+/// the custom event publisher
+// MARK: - CustomData
+struct CustomData: Codable {
+}
+
+// MARK: CustomData convenience initializers and mutators
+
+extension CustomData {
+    init(data: Data) throws {
+        self = try newJSONDecoder().decode(CustomData.self, from: data)
+    }
+
+    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
+        guard let data = json.data(using: encoding) else {
+            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
+        }
+        try self.init(data: data)
+    }
+
+    init(fromURL url: URL) throws {
+        try self.init(data: try Data(contentsOf: url))
+    }
+
+    func with(
+    ) -> CustomData {
+        return CustomData(
+        )
+    }
+
+    func jsonData() throws -> Data {
+        return try newJSONEncoder().encode(self)
+    }
+
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// A free-form object representing data specific to this event provided by
+/// Shopify. Refer to [standard events](#standard-events) for details on the
+/// payload available to each event
+// MARK: - PurpleData
+struct PurpleData: Codable {
```

Comment:
what is purple data?

Reply:
Not sure why it was added, I cannot find it in the original code. I'll remove it.

Reply:
It looks like that it refers to this:

```
/**
 * A free-form object representing data specific to this event provided by
 * Shopify. Refer to [standard events](#standard-events) for details on the
 * payload available to each event
 */
export type Data = Record<string, unknown>;
```

It is not being used, so I'll remove it.